resolver/addressable_resolver_test.go

```diff
@@ -628,6 +649,231 @@ Namespace: a lowercase RFC 1123 label must consist of lower case alphanumeric ch
+		dest            duckv1.Destination
+		addr            *unstructured.Unstructured
+		customResolvers []resolver.RefResolverFunc
+		wantAddress     string
+		wantErr         string
+	}{"address and addresses are not set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableNoAddresses(),
+		},
+		addr:    addressableNoAddresses(),
+		wantErr: fmt.Sprintf("address not set for %+v", addressableKnativeRef()),
+	}, "ref.address is set on destination and is present in target addressable addresses": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRefWithAddress(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddresses(),
+		},
+		wantAddress: *addressableKnativeRefWithAddress().Address,
+		addr:        addressableWithAddresses(),
+	}, "ref.address is set on destination and is NOT present in target addressable addresses": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRefWithAddress(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithDifferentAddresses(),
+		},
+		addr:    addressableWithDifferentAddresses(),
+		wantErr: fmt.Sprintf("address with name %q not found for %+v", *addressableKnativeRefWithAddress().Address, addressableKnativeRefWithAddress()),
+	}, "address and addresses are set but no destination.ref.address is set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddresses(),
+		},
+		addr:        addressableWithAddresses(),
+		wantAddress: addressableName,
+	}, "only address is set and no destination.ref.address is set": {
+		dest: duckv1.Destination{
+			Ref: addressableKnativeRef(),
+			URI: &apis.URL{
+				Path: "/foo",
+			},
+		},
+		objects: []runtime.Object{
+			addressableWithAddressOnly(),
+		},
+		addr:        addressableWithAddressOnly(),
+		wantAddress: addressableName,
+	}}
+
+	for n, tc := range tests {
+		t.Run(n, func(t *testing.T) {
+			ctx, _ := fakedynamicclient.With(context.Background(), scheme.Scheme, tc.objects...)
+			ctx = addressable.WithDuck(ctx)
+			r := resolver.NewURIResolverFromTracker(ctx, tracker.New(func(types.NamespacedName) {}, 0), tc.customResolvers...)
+
+			// Run it twice since this should be idempotent. URI Resolver should
+			// not modify the cache's copy.
+			_, _ = r.AddressableFromDestinationV1(ctx, tc.dest, tc.addr)
+			addr, gotErr := r.AddressableFromDestinationV1(ctx, tc.dest, tc.addr)
+
+			if gotErr != nil {
+				if tc.wantErr != "" {
+					if got, want := gotErr.Error(), tc.wantErr; got != want {
+						t.Errorf("Unexpected error (-want, +got) =\n%s", cmp.Diff(want, got))
+					}
+				} else {
+					t.Error("Unexpected error:", gotErr)
+				}
+			}
+			if got, want := *addr.Name, tc.wantAddress; got != want {
```

Comment:
`addr.Name` might be nil, so `*addr.Name` might panic, I'd suggest to make `wantAddress` of the same type as `addr` and compare `wantAddress` with `addr` to check the test result